### PR TITLE
xrt_core log with va_list

### PR DESCRIPTION
**Pull request by sarab96**
xrt_core log with va_list

### CODE REVIEW

File: src/runtime_src/core/common/message.cpp [hunk 1]
```diff
@@ -277,5 +278,21 @@ send(severity_level l, const char* tag, const char* msg)
...
   }
 }
+
+void
+send(severity_level l, const char* tag, const char* format, va_list args)
+{
+  static auto verbosity = xrt_core::config::get_verbosity();
+  if (l > (xrt_core::message::severity_level)verbosity) {
+    return;
+  }
```
Comment (stsoe) [inline src/runtime_src/core/common/message.cpp:288]: Remove the braces, leave blank line after return.

File: src/runtime_src/core/common/message.cpp [hunk 2]
```diff
@@ -277,5 +278,21 @@ send(severity_level l, const char* tag, const char* msg)
...
+  std::vector<char> msg_buff(MAX_LOGMSG_SIZE);
```
Comment (stsoe) [inline src/runtime_src/core/common/message.cpp:289]: Can we compute the required size?   Similar to how xclLogMsg does?

File: src/runtime_src/core/common/message.h [hunk 3]
```diff
@@ -22,6 +22,8 @@
 #include <cstdio>
 #include <vector>
 
+#define MAX_LOGMSG_SIZE 512
```
Comment (stsoe) [inline src/runtime_src/core/common/message.h:25]: Remove?

**Review by stsoe — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/common/message.cpp (hunk 1 above)
Comment (sarab96) [reply src/runtime_src/core/common/message.cpp]: done

File: src/runtime_src/core/common/message.cpp (hunk 2 above)
Comment (sarab96) [reply src/runtime_src/core/common/message.cpp]: done

File: src/runtime_src/core/common/message.h (hunk 3 above)
Comment (sarab96) [reply src/runtime_src/core/common/message.h]: done

File: src/runtime_src/core/common/message.cpp [hunk 4]
```diff
@@ -277,5 +279,40 @@ send(severity_level l, const char* tag, const char* msg)
...
+  if (len <= 0) {
+    //illegal arguments
+    std::string err_str = "ERROR: Illegal arguments in log format string. ";
+    err_str.append(std::string(format));
+    send(l, tag, err_str);
+    return;
+  }
+  ++len; //To include null terminator
+  std::vector<char> buf(len);
+  std::fill(buf.begin(), buf.end(), 0);
```
Comment (stsoe) [inline src/runtime_src/core/common/message.cpp:304]: How about `std::vector<char> buf(len, 0)`;

File: src/runtime_src/core/common/message.cpp [hunk 5]
```diff
@@ -277,5 +279,40 @@ send(severity_level l, const char* tag, const char* msg)
...
+  
+  va_list args_bak;
+  // vsnprintf will mutate va_list so back it up
+  va_copy(args_bak, args);
+  int len = std::vsnprintf(nullptr, 0, format, args_bak);
+  va_end(args_bak);
+  if (len <= 0) {
+    //illegal arguments
+    std::string err_str = "ERROR: Illegal arguments in log format string. ";
+    err_str.append(std::string(format));
```
Comment (stsoe) [inline src/runtime_src/core/common/message.cpp:298]: `err_str.append(format)`

File: src/runtime_src/core/common/message.cpp [hunk 6]
```diff
@@ -277,5 +279,40 @@ send(severity_level l, const char* tag, const char* msg)
...
+    return;
+  }
+  ++len; //To include null terminator
+  std::vector<char> buf(len);
+  std::fill(buf.begin(), buf.end(), 0);
+  len = std::vsnprintf(buf.data(), len, format, args);
+  if (len < 0) {
+    //error processing arguments
+    std::string err_str = "ERROR: When processing arguments in log format string. ";
+    err_str.append(std::string(format));
```
Comment (stsoe) [inline src/runtime_src/core/common/message.cpp:309]: err_str.append(format)

File: src/runtime_src/core/common/message.cpp [hunk 7]
```diff
@@ -277,5 +279,40 @@ send(severity_level l, const char* tag, const char* msg)
...
+  static auto verbosity = xrt_core::config::get_verbosity();
+  if (l > (xrt_core::message::severity_level)verbosity) 
+    return;
+  
+  va_list args_bak;
+  // vsnprintf will mutate va_list so back it up
+  va_copy(args_bak, args);
+  int len = std::vsnprintf(nullptr, 0, format, args_bak);
+  va_end(args_bak);
+  if (len <= 0) {
```
Comment (stsoe) [inline src/runtime_src/core/common/message.cpp:295]: Is 0 an error?  It's certainly a different condition and not an indication of invalid format string.

File: src/runtime_src/core/common/message.cpp [hunk 8]
```diff
@@ -277,5 +279,40 @@ send(severity_level l, const char* tag, const char* msg)
...
+    std::string err_str = "ERROR: Illegal arguments in log format string. ";
+    err_str.append(std::string(format));
+    send(l, tag, err_str);
+    return;
+  }
+  ++len; //To include null terminator
+  std::vector<char> buf(len);
+  std::fill(buf.begin(), buf.end(), 0);
+  len = std::vsnprintf(buf.data(), len, format, args);
+  if (len < 0) {
```
Comment (stsoe) [inline src/runtime_src/core/common/message.cpp:306]: Didn't we just check this?

**Review by stsoe — CHANGES_REQUESTED** [review]
I know it was copied from shim.cpp, but let's clean it up a bit.

File: src/runtime_src/core/common/message.cpp (hunk 4 above)
Comment (sarab96) [reply src/runtime_src/core/common/message.cpp]: this works

File: src/runtime_src/core/common/message.cpp (hunk 5 above)
Comment (sarab96) [reply src/runtime_src/core/common/message.cpp]: done

File: src/runtime_src/core/common/message.cpp (hunk 6 above)
Comment (sarab96) [reply src/runtime_src/core/common/message.cpp]: done

File: src/runtime_src/core/common/message.cpp (hunk 7 above)
Comment (sarab96) [reply src/runtime_src/core/common/message.cpp:295]: updated error msg. zero size log msg is unexpected and so reported as error with format string

File: src/runtime_src/core/common/message.cpp (hunk 8 above)
Comment (sarab96) [reply src/runtime_src/core/common/message.cpp]: changed

File: src/runtime_src/core/common/message.cpp [hunk 9]
```diff
@@ -277,5 +279,32 @@ send(severity_level l, const char* tag, const char* msg)
...
+    err_str.append(format);
+    send(l, tag, err_str);
+    return;
+  }
+  ++len; //To include null terminator
+  std::vector<char> buf(len, 0);
+  std::vsnprintf(buf.data(), len, format, args);
+  send(l, tag, buf.data());
+
+  return;
```
Comment (stsoe) [inline src/runtime_src/core/common/message.cpp:307]: Remove

**Review by stsoe — APPROVED** [review]
Just remove the return.

File: src/runtime_src/core/common/message.cpp (hunk 9 above)
Comment (sarab96) [reply src/runtime_src/core/common/message.cpp]: done

**Comment by sarab96** [comment]
retest this please

**Pull request merged**